pkg/cable/libreswan/libreswan.go

```diff
@@ -1,57 +1,368 @@
 package libreswan
 
 import (
+	"bufio"
 	"fmt"
+	"os"
+	"os/exec"
+	"strings"
+	"syscall"
 
-	v1 "github.com/submariner-io/submariner/pkg/apis/submariner.io/v1"
+	"github.com/kelseyhightower/envconfig"
+	"k8s.io/klog"
+
+	subv1 "github.com/submariner-io/submariner/pkg/apis/submariner.io/v1"
 	"github.com/submariner-io/submariner/pkg/cable"
 	"github.com/submariner-io/submariner/pkg/types"
+	"github.com/submariner-io/submariner/pkg/util"
 )
 
 const (
 	cableDriverName = "libreswan"
 )
 
 func init() {
-	cable.AddDriver(cableDriverName, NewLibreSwan)
+	cable.AddDriver(cableDriverName, NewLibreswan)
+}
+
+type libreswan struct {
+	secretKey string
+
+	debug   bool
+	logFile string
+
+	localEndpoint types.SubmarinerEndpoint
+
+	ipSecIKEPort  string
+	ipSecNATTPort string
+
+	// This tracks the requested connections
+	connections []subv1.Connection
 }
 
-type libreSwan struct {
+type specification struct {
+	PSK      string
+	Debug    bool
+	LogFile  string
+	IKEPort  string `default:"500"`
+	NATTPort string `default:"4500"`
 }
 
-// NewLibreSwan starts an IKE daemon using LibreSwan and configures it to manage Submariner's endpoints
-func NewLibreSwan(localEndpoint types.SubmarinerEndpoint, localCluster types.SubmarinerCluster) (cable.Driver, error) {
-	return &libreSwan{}, nil
+const defaultIKEPort = "500"
+const defaultNATTPort = "4500"
+const ipsecSpecEnvVarPrefix = "ce_ipsec"
+
+// NewLibreswan starts an IKE daemon using Libreswan and configures it to manage Submariner's endpoints
+func NewLibreswan(localEndpoint types.SubmarinerEndpoint, localCluster types.SubmarinerCluster) (cable.Driver, error) {
+	ipSecSpec := specification{}
+
+	err := envconfig.Process(ipsecSpecEnvVarPrefix, &ipSecSpec)
+	if err != nil {
+		return nil, fmt.Errorf("error processing environment config for %s: %v", ipsecSpecEnvVarPrefix, err)
+	}
+
+	return &libreswan{
+		secretKey:     ipSecSpec.PSK,
+		debug:         ipSecSpec.Debug,
+		logFile:       ipSecSpec.LogFile,
+		ipSecIKEPort:  ipSecSpec.IKEPort,
+		ipSecNATTPort: ipSecSpec.NATTPort,
+		localEndpoint: localEndpoint,
+		connections:   []subv1.Connection{},
+	}, nil
 }
 
 // GetName returns driver's name
-func (i *libreSwan) GetName() string {
+func (i *libreswan) GetName() string {
 	return cableDriverName
 }
 
 // Init initializes the driver with any state it needs.
-func (i *libreSwan) Init() error {
+func (i *libreswan) Init() error {
+	// Write the secrets file:
+	// %any %any : PSK "secret"
+	// TODO Check whether the file already exists
+	file, err := os.Create("/etc/ipsec.d/submariner.secrets")
+	if err != nil {
+		return fmt.Errorf("Error creating the secrets file: %v", err)
```

Comment:
nit: When returning an error message in an API, it's recommended to use small letter (i.e., instead of Error, please use error)

Reply:
Done, thanks for the reminder!